Add object bucket claim details to admin guide (#3882)
rook/rook: Documentation/ceph-object.md

kubectl -n rook-ceph get secret rook-ceph-object-user-my-store-my-user -o yaml | grep AccessKey | awk '{print $2}' | base64 --decode
kubectl -n rook-ceph get secret rook-ceph-object-user-my-store-my-user -o yaml | grep SecretKey | awk '{print $2}' | base64 --decode
```
## Create an Object Bucket Claim

> Using Object Bucket Claim(otherwise known as obc), a bucket will be provisioned from object store for the application. It can consume already existing buckets or else new bucket will be created. For this first a storage class need to be defined and then use it to create object bucket claim, all using prefered settings. See the [Object Bucket Claim CRD](ceph-object-bucket-claim-crd.md) for more detail on the settings available for a `CephObjectBucketClaim`.

Reviewer: - space after Claim before left paren.
- Capitalized "OBC".
- no need to include "otherwise known as" text

Author: Done

> Using Object Bucket Claim(otherwise known as obc), a bucket will be provisioned from object store for the application. It can consume already existing buckets or else new bucket will be created. For this first a storage class need to be defined and then use it to create object bucket claim, all using prefered settings. See the [Object Bucket Claim CRD](ceph-object-bucket-claim-crd.md) for more detail on the settings available for a `CephObjectBucketClaim`.

Reviewer 2: "Using an Object Bucket Claim (OBC), ..."

Author: Done

```yaml
apiVersion: storage.k8s.io/v1

> kind: StorageClass

Reviewer 2: Need to mention that this storage class is for new bucket provisioning -- not existing buckets.

Author: Done

metadata:
   name: rook-ceph-bucket
provisioner: ceph.rook.io/bucket
reclaimPolicy: Delete
parameters:
  objectStoreName: my-store
  objectStoreNamespace: rook-ceph
  region: us-east-1
```
```bash

> # Create the Storage class for bucket provisioner

Reviewer 2: need to mention this creating step is done by an administrator.

Author: Done

kubectl create -f storageclass-bucket.yaml
```
```yaml
apiVersion: objectbucket.io/v1alpha1
kind: ObjectBucketClaim
metadata:
  name: ceph-bucket
spec:

>   bucketName: ceph-bkt

Reviewer 2: This is the atypical use. It will be more common to specify `generateBucketName: ceph-prefix`

Author: Done

  storageClassName: rook-ceph-bucket
```

> After creating the OBC, a secret and config map will be configure to access the object store with similar name. The creditionals can be fetched using following commands :

Reviewer 2: Change to:

"After creating the OBC, a secret and configmap are created with the same name as the OBC and in the same namespace. The secret contains credentials used by the application pod to access the bucket. The configmap contains bucket endpoint information and is also consumed by the pod.

The following commands extract key pieces of information from the secret and configmap:"

Author: Done

```bash

> kubectl get cm ceph-bucket -o yaml | grep BUCKET_HOST | awk '{print $2}'

Reviewer 2: Add `-n <namespace>` here or add a comment indicating to use `-n <namespace>`. 
Lack of namespace in these commands implies either 1) the default ns is being used (which won't happen in production clusters), or 2) kubeconfig context is set to the user's namespace (likely but needs to be explicit).

Author: Done

kubectl -n rook-ceph get secret rook-ceph-object-user-my-store-my-user -o yaml | grep AccessKey | awk '{print $2}' | base64 --decode
kubectl -n rook-ceph get secret rook-ceph-object-user-my-store-my-user -o yaml | grep SecretKey | awk '{print $2}' | base64 --decode
```

> ## Create an Object Bucket Claim

Reviewer 2: How much do you want to repeat in this doc vs. reference links to the first doc (above)?

> Creating an Object Bucket Claim (OBC) causes the Rook-Ceph bucket provisioner to create a new bucket or to grant access to an existing bucket. The OBC references a storage class, set up by the administrator, which defines whether or not a new bucket is created, and the retention policy for new buckets. See the [Object Bucket Claim](ceph-object-bucket-claim.md) for more detail on the `CephObjectBucketClaim`.

Reviewer: Some minor grammar things, and the sentence in the middle is very complex with too many comma-separated clauses to be understood easily.

```suggestion
Creating an Object Bucket Claim (OBC) causes the Rook-Ceph bucket provisioner to create a new bucket or to grant access to an existing bucket. When the OBC does not reference an existing storage class set up by an administrator, a new bucket is created with the specified retention policy. See the [Object Bucket Claim documentation](ceph-object-bucket-claim.md) for more detail on `CephObjectBucketClaims`.
```

> Creating an Object Bucket Claim (OBC) causes the Rook-Ceph bucket provisioner to create a new bucket or to grant access to an existing bucket. When the OBC does not references an existing storage class set up by an administrator, a new bucket is created with the specified retention policy. See the [Object Bucket Claim Documentation](ceph-object-bucket-claim.md) for more detail on the `CephObjectBucketClaims`.

Reviewer 2: Try rewording :

_Creating an Object Bucket Claim (OBC) causes the Rook-Ceph bucket provisioner to create a new bucket or to grant access to an existing bucket. OBCs reference a Storage Class which is created by a cluster administrator. The Storage Class defines the object storage system, the bucket retention policy, and, when granting access to existing (brownfield) buckets, the name of the bucket. If a new bucket is desired then the OBC contains the bucket name or a prefix to be used for a random bucket name. See the [Object Bucket Claim Documentation](ceph-object-bucket-claim.md) for more detail on the `CephObjectBucketClaims`._

> ```yaml

Reviewer 2: What is the reason to repeat the yamls here? Additionally, this order of the yamls here is different from the other doc.

Reviewer 2: I would delete L113 - L147 unless there is info unique to this doc.

apiVersion: storage.k8s.io/v1

> kind: StorageClass

Reviewer 2: I still don't understand why the first thing shown is a storage class? See my suggested order above?

Author: Done

metadata:
   name: rook-ceph-bucket
provisioner: ceph.rook.io/bucket
reclaimPolicy: Delete
parameters:
  objectStoreName: my-store
  objectStoreNamespace: rook-ceph
  region: us-east-1
```
```bash

> # Above Storage class for new bucket provisioning, below command need to executed by administrator

Reviewer 2: Change to:
"The storage class defines new bucket provisioning and is created by the admin"

Author: Done

kubectl create -f storageclass-bucket.yaml
```
```yaml
apiVersion: objectbucket.io/v1alpha1
kind: ObjectBucketClaim
metadata:
  name: ceph-bucket
spec:
  generateBucketName: ceph-bkt
  storageClassName: rook-ceph-bucket
```

> After creating the OBC, a secret and configmap are created with the same name as the OBC and in the same namespace. The secret contains credentials used by the application pod to access the bucket. The configmap contains bucket endpoint information and is also consumed by the pod.

Reviewer: ```suggestion
After creating the OBC, a secret and ConfigMap are created with the same name as the OBC and in the same namespace. The secret contains credentials used by the application pod to access the bucket. The configmap contains bucket endpoint information and is also consumed by the pod.
```

Reviewer: Will the user ever have to consume this information themselves manually, or will this be done via Flex/CSI driver automatically when making a volume claim?

Reviewer 2: > Will the user ever have to consume this information themselves manually,
Yes. App pods need to consume the generated secret and configmap in order to obtain bucket credentials and endpoint.

> After creating the OBC, a secret and configmap are created with the same name as the OBC and in the same namespace. The secret contains credentials used by the application pod to access the bucket. The configmap contains bucket endpoint information and is also consumed by the pod.

Reviewer 2: Change slightly to:
"After creating the OBC and successful bucket provisioning, a secret and configmap are automatically created ..."

> The following commands extract key pieces of information from the secret and configmap:"

Reviewer 2: add "... from the generated secret and..."  I want to keep reminding the reader that _they_ don't create the secret or CM.

```bash
#config map, secret, OBC will part of default if no specific name space mentioned
kubectl -n default get cm ceph-bucket -o yaml | grep BUCKET_HOST | awk '{print $2}'
kubectl -n default get secret ceph-bucket -o yaml | grep AWS_ACCESS_KEY_ID | awk '{print $2}' | base64 --decode
kubectl -n default get secret ceph-bucket -o yaml | grep AWS_SECRET_ACCESS_KEY | awk '{print $2}' | base64 --decode

> ```

Reviewer 3: One more piece that feels missing is an example of how to use these variables from a client pod that is accessing the bucket. A full example pod would be nice in the examples folder, but here in the docs we might at least put a yaml snippet of what those settings look like in the pod spec.

kubectl -n rook-ceph get secret rook-ceph-object-user-my-store-my-user -o yaml | grep AccessKey | awk '{print $2}' | base64 --decode
kubectl -n rook-ceph get secret rook-ceph-object-user-my-store-my-user -o yaml | grep SecretKey | awk '{print $2}' | base64 --decode
```
## Create an Object Bucket Claim

> Creating an Object Bucket Claim (OBC) causes the Rook-Ceph bucket provisioner to create a new bucket or to grant access to an existing bucket. OBCs reference a Storage Class which is created by a cluster administrator. The Storage Class defines the object storage system, the bucket retention policy, and, when granting access to existing (brownfield) buckets, the name of the bucket. If a new bucket is desired then the OBC contains the bucket name or a prefix used for a random bucket name. See the [Object Bucket Claim Documentation](ceph-object-bucket-claim.md) for more detail on the `CephObjectBucketClaims`.

Author: This may need to change entire flow of the entire doc, even for accessing the object store part as well. Can we take altogether as different patch

> Creating an Object Bucket Claim (OBC) causes the Rook-Ceph bucket provisioner to create a new bucket or to grant access to an existing bucket. OBCs reference a Storage Class which is created by a cluster administrator. The Storage Class defines the object storage system, the bucket retention policy, and, when granting access to existing (brownfield) buckets, the name of the bucket. If a new bucket is desired then the OBC contains the bucket name or a prefix used for a random bucket name. See the [Object Bucket Claim Documentation](ceph-object-bucket-claim.md) for more detail on the `CephObjectBucketClaims`.

Reviewer 3: I think this is critical to get the new flow clear. Instead of trying to explain this in detail, I made a pass at these changes and the commit is pushed [here](https://github.com/travisn/rook/commit/ba3d9a8df41b8a0e90fd7961f04868933f8619a1). Want to cherry pick this commit and include it in the PR?
